stout/task.h

```diff
@@ -72,223 +84,298 @@ struct HeapTask {
+  template <typename K_, typename Value_, typename... Args_>
+  struct Continuation {
+    template <typename... Args>
+    void Start(Args&&...) {
+      std::apply(
+          [&](auto&&... args) {
+            dispatch_(
+                Action::Start,
+                std::nullopt,
+                std::forward<decltype(args)>(args)...,
+                e_,
+                *interrupt_,
+                [this](auto&&... args) {
+                  k_.Start(std::forward<decltype(args)>(args)...);
+                },
+                [this](std::exception_ptr e) {
+                  k_.Fail(std::move(e));
+                },
+                [this]() {
+                  k_.Stop();
+                });
+          },
+          std::move(args_));
+    }
 
-  template <typename F>
-  TaskWith(Args_... args, F f)
-    : args_(std::move(args)...) {
-    // static_assert(
-    //     !IsContinuation<F>::value,
-    //     "'Task' expects a callable "
-    //     "NOT an eventual continuation");
-
-    static_assert(
-        sizeof...(args) > 0 || std::is_invocable_v<F>,
-        "'Task' expects a callable that "
-        "takes no arguments");
-
-    static_assert(
-        sizeof...(args) == 0 || std::is_invocable_v<F, Args_...>,
-        "'Task' expects a callable that "
-        "takes the arguments specified");
-
-    static_assert(
-        sizeof(f) <= sizeof(void*),
-        "'Task' expects a callable that "
-        "can be captured in a 'Callback'");
-
-    using E = decltype(f(std::move(args)...));
-
-    // static_assert(
-    //     IsContinuation<E>::value,
-    //     "'Task' expects a callable that returns "
-    //     "an eventual continuation");
-
-    using Value = typename E::template ValueFrom<void>;
-
-    static_assert(
-        std::is_convertible_v<Value, Value_>,
-        "eventual result type can not be converted into type of 'Task'");
-
-    start_ = [f = std::move(f)](
-                 Args_&&... args,
-                 std::unique_ptr<void, Callback<void*>>& e_,
-                 Interrupt& interrupt,
-                 Callback<Value>&& start,
-                 Callback<std::exception_ptr>&& fail,
-                 Callback<>&& stop) {
-      if (!e_) {
-        e_ = std::unique_ptr<void, Callback<void*>>(
-            // TODO(benh): pass the args to 'Start()' instead so that
-            // they don't have to get copied more than once in the
-            // event that the eventual returned from 'f' copies them?
-            new detail::HeapTask<E>(f(std::move(args)...)),
-            [](void* e) {
-              delete static_cast<detail::HeapTask<E>*>(e);
-            });
-      }
-
-      auto* e = static_cast<detail::HeapTask<E>*>(e_.get());
-
-      e->Start(interrupt, std::move(start), std::move(fail), std::move(stop));
-    };
-  }
+    template <typename... Args>
+    void Fail(Args&&... args) {
+      auto exception = std::make_exception_ptr(
+          std::forward<decltype(args)>(args)...);
```

Comment:
I'm realizing that we don't handle the case when `args` are empty here or in `terminal.h`. Can you please handle that case here by just using a `std::runtime_error` and then also do the same in `terminal.h`? Thanks!